### PR TITLE
TS-4506 Don't remove Expires/Last-Modifed on 304 response

### PR DESCRIPTION
**Pull request by zwoop**
When generating a 304 page, we used to remove Expires and Last-Modified
always. This change makes sure we never touch Expires, and only remove
the Last-Modified header if there is an ETag in the response. This is per
the specs in

    https://tools.ietf.org/html/rfc7232#page-18

### CODE REVIEW

**Comment by zwoop** [comment]
@bryancall @masaori335 Please review.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/358/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/252/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/261/ for details.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/366/ for details.

File: proxy/http/HttpTransact.cc [hunk 1]
```diff
@@ -8199,13 +8199,15 @@ HttpTransact::build_error_response(State *s, HTTPStatus status_code, const char
     if (has_ua_msie)
       s->hdr_info.client_response.value_set(MIME_FIELD_PROXY_CONNECTION, MIME_LEN_PROXY_CONNECTION, "close", 5);
   }
-  // Add a bunch of headers to make sure that caches between
-  // the Traffic Server and the client do not cache the error
-  // page.
+  // Make sure that caches between the Traffic Server and the client do not cache the error.
+  // ToDo: Not sure this matches expectation in RFC 7223.
```
Comment (masaori335) [inline proxy/http/HttpTransact.cc]: Looks a typ. RFC 7232 is correct?

**Comment by masaori335** [comment]
Do we need some changes in `HttpTransact::build_response()` too?
https://github.com/apache/trafficserver/blob/master/proxy/http/HttpTransact.cc#L7945-L7975.

**Comment by zwoop** [comment]
I updated the PR as per William's post on the Jira. I had a bunch of typos and errors here, I hope this fixes those issues.

File: proxy/http/HttpTransact.cc (hunk 1 above)
Comment (zwoop) [inline proxy/http/HttpTransact.cc]: Well, I'm open to suggestion here. It seems, as per the RFC, that we probably shouldn't modify the Cache-Control header here? That seems inline with the code build_response() function too, right ?

**Comment by zwoop** [comment]
@masaori335 HttpTransact::build_response() seems correct, doesn't it? But, seeing that, also makes me think that the place where we modify Cache-Control: (see your comment above) might actually be wrong?

Wdyt?

**Comment by wbardwel** [comment]
I assumed that the Cache-Control: no-store stuff was because that function is not mostly used for 304s...it is mostly used for 5xx errors, it might need to branch to handle different status codes differently.

build_response() seems to no try to pass along Last-Modified headers, only ETag which seems slightly wrong...(but doesn't have the other issues that this is about)

**Comment by zwoop** [comment]
So, do we want to change it such that this isn't used for 304's? Or do we want to make it conditional in here for 304's? Maybe build_error_response() should avoid setting or modifying headers at all?

**Comment by zwoop** [comment]
Another option would be to just not remove the Last-Modified here, but unclear what exactly implications that has. I guess I really ought to look through all possible status codes, and see if the current logic makes sense everywhere :-/. Why can't anything be simple?

**Comment by mlibbey** [comment]
The RFC portion didn't mandate removing Last-Modified. ("Since the goal of a 304 response is to minimize information transfer when the recipient already has one or more cached representations, a sender SHOULD NOT generate representation metadata other than the above listed fields unless said metadata exists for the purpose of guiding cache updates (e.g., Last-Modified might be useful if the response does not have an ETag field)." I've seen one case where others were expecting it to be their regardless of the ETag (because another large commercial CDN does it that way).

File: proxy/http/HttpTransact.cc (hunk 1 above)
Comment (bryancall) [inline proxy/http/HttpTransact.cc]: It should be RFC7232 not RFC7223, like @masaori335 said.

File: proxy/http/HttpTransact.cc [hunk 2]
```diff
@@ -8199,13 +8199,15 @@ HttpTransact::build_error_response(State *s, HTTPStatus status_code, const char
...
   s->hdr_info.client_response.value_set(MIME_FIELD_CACHE_CONTROL, MIME_LEN_CACHE_CONTROL, "no-store", 8);
-  // Make sure there are no Expires and Last-Modified headers.
-  s->hdr_info.client_response.field_delete(MIME_FIELD_EXPIRES, MIME_LEN_EXPIRES);
-  s->hdr_info.client_response.field_delete(MIME_FIELD_LAST_MODIFIED, MIME_LEN_LAST_MODIFIED);
+
+  // Remove the Last-Modified header, but only if there is an ETag header. See sec 4.1 in
+  //     https://tools.ietf.org/html/rfc7232#page-18
```
Comment (bryancall) [inline proxy/http/HttpTransact.cc]: I think it would be better to link to the section and not the page:
https://tools.ietf.org/html/rfc7232#section-4.1

**Comment by bryancall** [comment]
It should be conditional on the 304 in the build_error_response and also added to build_response.  Maybe pull the logic into a method and call it from both.

**Comment by zwoop** [comment]
Yeah, I need to work on this some more, not sure exactly yet what path to take yet. We have some options :). I'd like to keep this open, so we retain the feedback.

**Pull request closed without merging**